### DISCUSSION TITLE
Method inlined on Android 13/Tiramisu (API 33)

### DISCUSSION THREAD

**Discussion by vegedreamgagnoa** [LSPosed/LSPlant#5] [Q&A] [answered]
Tried LSPlant to hook android.app.Activity#getSystemService as a sample, the hook result returns true, but the callback has never been called. Why? trying to deoptimize not work. **It works on Android 6 ~ 12** though.

**Comment by yujincheng08** [comment]
Have you tried to invoke `android.app.Activity#getSystemService` by reflection?

**Comment by vegedreamgagnoa** [comment]
If I am not mistaken it requires the object/instance of `android.app.Activity`, isn't it? Which usually we don't have access to that (we only have access to the Activity we defined). It's quite an hassle especially if we don't have access to the owner object.
Looking at the source, I am still not understand how this is different to/surpasses Pine or Whale - where with both frameworks I can easily hook that method.

**Comment by yujincheng08** [comment]
What I mean is just for testing...

**Comment by vegedreamgagnoa** [comment]
Oh, sorry - yeah, it works if I tried to invoke the backup using reflection.

**Comment by yujincheng08** [comment]
Not the backup.

What I mean is:
Hook, invoke the original method by reflection and see if the hook is invoked.

**Comment by vegedreamgagnoa** [comment]
Nope, not working - only works for Activity that I have defined, as I said.

**Comment by yujincheng08** [comment]
So reflection (on your own activity) can trigger hook, right?

**Comment by vegedreamgagnoa** [comment]
Yes! :)

**Comment by vegedreamgagnoa** [comment]
I am still trying to figure out why it's not working only for Tiramisu...

**Comment by yujincheng08** [comment, marked as answer]
Ok, then it's the inline problem. Tiramisu has increased its inline threshold so it's not surprising.
See https://github.com/LSPosed/LSPosed/commit/ef1439a3f36dc57ad7b72161a759e5960b1ce91c, `systemMain` is also inlined...

As the document has said, you should deoptimize the caller to make it work. Note that it's not deoptimizing the `getSystemService` but the method which invokes `getSystemService`.

For example, there's a piece of code like this:

```
class X {
fun A() {
xxActivity.getSystemService("activity")
}
}
```

You should `deoptmize(X::A)` rather than `deoptmize(Activity::getSystemService)`

**Comment by vegedreamgagnoa** [comment]
Cool, thanks for the hint. Can you also suggest how to find the caller of the method at runtime?

**Comment by yujincheng08** [comment]
> Cool, thanks for the hint. Can you also suggest how to find the caller of the method at runtime?

It's possible but slow. I recommend you to hard code the callers. But if you are interesting, you can take a look at our experimental API for it at:

https://github.com/LSPosed/DexBuilder/blob/master/include/dex_helper.h

It can search all methods that invoke a specific method.

And a JVM binding is here: https://github.com/yujincheng08/BiliRoaming/blob/master/app/src/main/jni/biliroaming.cc


These APIs are still for experimental usage. We will publish it when it matures.

**Comment by vegedreamgagnoa** [comment]
This is working: https://github.com/LSPosed/LSPosed/blob/master/core/src/main/cpp/external/yahfa/src/HookMain.cpp

**Comment by yujincheng08** [comment]
are you hooking a normal app? LSPosed has disabled inline systemwisely so it may work.

**Comment by vegedreamgagnoa** [comment]
Yeah, I didn't build any Xposed/LSPosed module, just include this library and try on a normal app.
It's still not working as I had hoped, invoking the original using reflection might be good but without the owner object, you can't really do anything...

**Comment by vegedreamgagnoa** [comment]
Why this closed? We still have a  discussion! ))

**Comment by yujincheng08** [comment]
> Not the backup.

**Comment by vegedreamgagnoa** [comment]
woops, sorry, I mean the original. edited :)

**Comment by yujincheng08** [comment]
> you should deoptimize the caller to make it work.

**Comment by yujincheng08** [comment]
And. The inline problem affects all hook frameworks, including pine, epic, sandhook, yahfa, etc.

**Comment by vegedreamgagnoa** [comment]
@yujincheng08 I found this function getInstanceOfClasses on VMDebug class on Android P and higher, might be able to use it to locate the instances/callers?

**Comment by yujincheng08** [comment]
Reflection is just for tests. If a hook can be invoked by reflection but not by other callers, it's likely inlined.

And the solution is never to find the instances and invoke them by reflection. The solution is to avoid the inline, like deoptimizing all possible callers or recompiling the app with flags that disable the inline.

**Comment by yujincheng08** [comment]
> Why this closed? We still have a discussion! ))

It's not a bug. Maybe we should move it to the discussion.

**Comment by vegedreamgagnoa** [comment]
Which Android versions that have this inlining problem? Android R, S, and Tiramisu? or just Android S and Tiramisu?

**Reply by yujincheng08** [reply]
Nearly all.

**Reply by yujincheng08** [reply]
`inline-max-code-units` appears in Android 6. So at least Android 6 has the inline problem.

**Comment by yujincheng08** [comment]
Read https://github.com/asLody/SandHook/blob/master/doc/doc.md#inline-%E5%A4%84%E7%90%86 about the inline problem and how to deal with it.

**Comment by vegedreamgagnoa** [comment]
@yujincheng08 Seems like everything works correctly. I tried different method, this time, it's `android.app.Activity#dispatchTouchEvent`. The callback has been called successfully, and I returned true in the callback, but I couldn't tap on anything on my app. Maybe I need to call the super.dispatchTouchEvent (the original method) on the callback? How do I get this object/instance so I can invoke that original method?

**Reply by vegedreamgagnoa** [reply]
I am thinking of using VMDebug#getInstanceOfClasses, and send that to my hooker object as the owner object - from there I can invoke the original method.

**Reply by yujincheng08** [reply]
args[0] is the `this` object for non-static method.

**Reply by vegedreamgagnoa** [reply]
Yeay! It works!

**Reply by vegedreamgagnoa** [reply]
Thank you very much @yujincheng08 sorry to bother you too much
It was my mistake on the end.

**Reply by yujincheng08** [reply]
No worries.

Maybe the doc is somehow confusing to users, so PRs to improve the doc is always welcoming, especially from those real users.

**Reply by vegedreamgagnoa** [reply]
Sure, if there's anything I come with I will try to send a PR :)
By the way, why didn't support Android 5/5.1? Is there anything different?
The last time I checked it's because there's no java/lang/Executable or AbstractMethod in Android 5/5.1.
Is there any alternative hooking mechanism that we can use/you can suggest?

**Reply by yujincheng08** [reply]
There's AbstractMethod in Android 5.

The thing is.
In Android 6, the art method field is in long type, which is a pointer to the art method native object.
However, in Android 5, the art method field is a java object, and the logic to modify entrypoint or so is totally different.

**Reply by vegedreamgagnoa** [reply]
@yujincheng08 Is it possible for us to know the called/fired method from the callback?
From `public Object callback_method(Object[] args)`, how can I know the method being called on this callback?

**Reply by vegedreamgagnoa** [reply]
Never mind - I know how :)

**Reply by yujincheng08** [reply]
store in the hooker object.
you can store any context and info you need in it.
